content/articles/integrate-ml-to-flask-api-and-deploy-to-heroku/index.md

```diff
@@ -0,0 +1,336 @@
+
+```python
+# pickle the training set
+import pickle
+filename_X = 'X_train.sav'
+pickle.dump(X_train, open(filename_X, 'wb'))
+filename = 'model.sav'
+pickle.dump(model, open(filename, 'wb'))
+```
+
+Running the code above creates two new files in your current directory: `X_train.sav` and `model.sav`.
+
+
+### Integrating ML model to a Flask-RESTful API
+
+Using a text editor (VS Code), create a new file and rename it to `api.py.` 
+We create a virtual environment to isolate the libraries we will be using for this project. The virtual environment will not change the modules installed on our system. Instead, it will create a unique environment where we can install only the libraries we need for our project. Thus, it helps us avoid conflict between the libraries installed in our system and the libraries we will be using for our project. First, we install `virtualenv`. Then, we use `virtualenv` to build separate Python environments.
+
+```bash
+pip install virtualenv
+```
+
+We create a virtual environment and call it venv. It adds a new folder venv to our directory.
+
+```bash
+virtualenv venv
+```
+
+To activate the virtual environment on Windows, run the command below.
+
+```bash
+venv\Scripts\activate.bat 
+```
+
+Once we activate the virtual environment, we install all the libraries we need to run our API. The `pickle` module comes by default with python. So we do not need to install it. Next, we installed the version of the scikit-learn library we used to build the ML model in our Jupyter Notebook. To check for the version, run the code below in your notebook.
+
+```python
+import sklearn
+
+print('scikit-learn version is', sklearn.__version__)
+```
+
+**Output:**
+```
+scikit-learn version is 0.24.2
+```
+
+We install this version of scikit-learn in our virtual environment.
+
+```bash
+pip install flask
+pip install flask-restful
+pip install numpy
+pip install scikit-learn==0.24.2
+```
+
+Now we can build our RESTful API. We start by importing the libraries and instantiating the Flask RESTful API in `api.py`.
+
+```python
+from flask import Flask, request
+from flask_restful import Api, Resource
+import numpy as np
+import pickle
+from sklearn.preprocessing import StandardScaler
+
+# instantiate Flask Rest Api
+app = Flask(__name__)
+api = Api(app)
+```
+
+The Flask-RESTful library in Flask enables us to build REST APIs with ease. It strictly follows the REST API standard. First, let us load our pickled model and training set.
+
+```python
+# load the pickled model and X_train
+X_train = pickle.load(open('X_train.sav', 'rb'))
+model = pickle.load(open('model.sav', 'rb'))
+```
+
+We feature scale the training set. We will scale the new patient's data later before making predictions.

```

Comment:
What do you mean by feature scale the training set?

Reply:
I mean we apply standardization to the training set.